### PR TITLE
[COMGR] Fixes for HIP PCH

### LINKED ISSUE #624: [Comgr] Unknown type name occurred while compiling kernel
I configured miopen-2.8.0 with `-DMIOPEN_USE_COMGR=1` and compiled it. Then I used `./MIOpenDriver conv --batchsize 16 --in_channels 64 --out_channels 256 --iter 1 -V 0` to run `MIOpenDriver`. Eventually, I got an error while compiling kernel as follows:

<details><summary><b><i>Console log:</i></b></summary>

```
1387 MIOpen(HIP)  38.756: Info2 [SetIsaName] amdgcn-amd-amdhsa--gfx906+sram-ecc
1387 MIOpen(HIP)  92.296: Error [Do] 'amd_comgr_do_action(kind, handle, in.GetHandle(), out.GetHandle())' AMD_COMGR_ACTION_COMPILE_SOURCE_TO_BC: ERROR (1)
1387 MIOpen(HIP)   5.282: Error [BuildHip] comgr status = ERROR (1)
1387 MIOpen(HIP)   0.025: Warning [BuildHip] In file included from /tmp/comgr-0c6a05/input/gridwise_convolution_implicit_gemm_v4r1_nchw_kcyx_nkhw_lds_double_buffer.cpp:1:
In file included from /tmp/comgr-0c6a05/include/common_header.hpp:4:
In file included from /tmp/comgr-0c6a05/include/config.hpp:8:
/tmp/comgr-0c6a05/include/bfloat16_dev.hpp:41:5: error: unknown type name 'uint'; did you mean 'int'?
    uint u32;
    ^~~~
    int
/tmp/comgr-0c6a05/include/bfloat16_dev.hpp:42:5: error: unknown type name 'ushort2'; did you mean 'short'?
    ushort2 ushortx2;
    ^~~~~~~
    short
/tmp/comgr-0c6a05/include/bfloat16_dev.hpp:47:5: error: unknown type name 'ushort'; did you mean 'short'?
    ushort ushortvec[2];
    ^~~~~~
    short
/tmp/comgr-0c6a05/include/bfloat16_dev.hpp:52:45: error: unknown type name 'ushort'; did you mean 'short'?
EXECUTION_SPECIFIER float bfloat16_to_float(ushort src_val)
                                            ^~~~~~
                                            short
/tmp/comgr-0c6a05/include/bfloat16_dev.hpp:57:27: error: use of undeclared identifier 'make_ushort2'
    target_val.ushortx2 = make_ushort2(0, src_val);
                          ^
/tmp/comgr-0c6a05/include/bfloat16_dev.hpp:65:21: error: unknown type name 'ushort'; did you mean 'short'?
EXECUTION_SPECIFIER ushort float_to_bfloat16(float src_val)
                    ^~~~~~
                    short
In file included from /tmp/comgr-0c6a05/input/gridwise_convolution_implicit_gemm_v4r1_nchw_kcyx_nkhw_lds_double_buffer.cpp:1:
In file included from /tmp/comgr-0c6a05/include/common_header.hpp:4:
/tmp/comgr-0c6a05/include/config.hpp:113:17: error: unknown type name 'int32_t'
using index_t = int32_t;
                ^
/tmp/comgr-0c6a05/include/config.hpp:117:9: error: unknown type name 'int32_t'
typedef int32_t int32x4_t __attribute__((ext_vector_type(4)));
        ^
In file included from /tmp/comgr-0c6a05/input/gridwise_convolution_implicit_gemm_v4r1_nchw_kcyx_nkhw_lds_double_buffer.cpp:1:
In file included from /tmp/comgr-0c6a05/include/common_header.hpp:5:
/tmp/comgr-0c6a05/include/utility.hpp:8:12: error: unknown type name 'index_t'
__device__ index_t get_thread_local_1d_id() { return threadIdx.x; }
           ^
/tmp/comgr-0c6a05/include/utility.hpp:8:54: error: use of undeclared identifier 'threadIdx'
__device__ index_t get_thread_local_1d_id() { return threadIdx.x; }
                                                     ^
/tmp/comgr-0c6a05/include/utility.hpp:10:12: error: unknown type name 'index_t'
__device__ index_t get_block_1d_id() { return blockIdx.x; }
           ^
/tmp/comgr-0c6a05/include/utility.hpp:10:47: error: use of undeclared identifier 'blockIdx'
__device__ index_t get_block_1d_id() { return blockIdx.x; }
                                              ^
In file included from /tmp/comgr-0c6a05/input/gridwise_convolution_implicit_gemm_v4r1_nchw_kcyx_nkhw_lds_double_buffer.cpp:1:
In file included from /tmp/comgr-0c6a05/include/common_header.hpp:7:
/tmp/comgr-0c6a05/include/number.hpp:8:11: error: unknown type name 'index_t'
template <index_t N>
          ^
/tmp/comgr-0c6a05/include/number.hpp:9:34: error: use of undeclared identifier 'index_t'
using Number = integral_constant<index_t, N>;
                                 ^
/tmp/comgr-0c6a05/include/number.hpp:11:11: error: unknown type name 'index_t'
template <index_t X, index_t Y>
          ^
/tmp/comgr-0c6a05/include/number.hpp:11:22: error: unknown type name 'index_t'
template <index_t X, index_t Y>
                     ^
/tmp/comgr-0c6a05/include/number.hpp:12:46: error: no template named 'Number'
__host__ __device__ constexpr auto operator+(Number<X>, Number<Y>)
                                             ^
/tmp/comgr-0c6a05/include/number.hpp:12:57: error: no template named 'Number'
__host__ __device__ constexpr auto operator+(Number<X>, Number<Y>)
                                                        ^
/tmp/comgr-0c6a05/include/number.hpp:14:12: error: use of undeclared identifier 'Number'
    return Number<X + Y>{};
           ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated when compiling for gfx906.

terminate called after throwing an instance of 'miopen::Exception'
  what():  /home/wenjh/source_code/miopen_gitlab_sugon/src/hipoc/hipoc_program.cpp:285: Code object build failed. Source: gridwise_convolution_implicit_gemm_v4r1_nchw_kcyx_nkhw_lds_double_buffer.cpp
Aborted (core dumped)
```
</details>

Is my environment configured wrong?

### PR DESCRIPTION
**Pull request by atamazov**
Resolves issues related to SWDEV-257056:

- Fixed: Observed `MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS` in `handle_test`.
- Fixed: Observed `MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS` in reduction_operator.hpp (fixes `reduce_test`).
- Fixed: Observed `MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS` in ConvAsmImplicitGemmV4R1DynamicWrw.
- Fixed: `handle_test` for the case when HIP kernels disabled.
- Fixed: Incorrect `#include` in kernel source.
- Added: Workaround in `handle_test` for macros missing from HIP PCH in ROCm, 3.10 RC3
  - Macro: `WORKAROUND_SWDEV_257056_PCH_MISSING_MACROS`.
  - The workaround is enabled by default.
  - It will be disabled automatically as soon as ROCm version exceeds 3.10.x
- Added feature: Introduced env var control `MIOPEN_DEBUG_COMGR_HIP_PCH_ENFORCE` that allows overriding of HIP PCH feature auto-detection.
- Added: Workaround for incorrectly reported support for HIP PCH in ROCm 3.9
  - Macro: `WORKAROUND_SWDEV_257056_PCH_INCORRECTLY_REPORTED`.
  - Fixes #624.

### CODE REVIEW

**Comment by daniellowell** [comment]
@fronteer @shaojiewang @JehandadKhan You are all assigned to review this urgent PR, please do so.

File: src/comgr.cpp [hunk 1]
```diff
@@ -217,6 +219,28 @@ static void RemoveOptionsUnwanted(OptionList& list)
...
+static bool IsPchEnabled()
+{
+    if(miopen::IsEnabled(MIOPEN_DEBUG_COMGR_HIP_PCH_ENFORCE{}))
+        return true;
+    if(miopen::IsDisabled(MIOPEN_DEBUG_COMGR_HIP_PCH_ENFORCE{}))
+        return false;
+    return HIP_SUPPORTS_PCH;
+}
+
+static std::string GetPchEnableStatus()
```
Comment (daniellowell) [inline src/comgr.cpp:231]: I feel like this could be useful elsewhere and not just logging.

Comment (atamazov) [reply src/comgr.cpp:231]: Only here so far.

**Review by JehandadKhan — APPROVED** [review]

**Comment by atamazov** [comment]
Description updated.

**Review by daniellowell — APPROVED** [review]

**Pull request merged**